[P118] Moving Itho plugin the main repo

svollebregt opened the pull request:
Moving P145 (Itho fan remote) to the main repo as P118. Moved into Testing_D which compiles on my side in Atom.

Related to issue #3746

svollebregt committed:
Adding Itho library to main reposity

Moving P145 from the playgorund to the main repo as P118.

svollebregt committed:
Revert "Adding Itho library to main reposity"

This reverts commit 1cfdd9e8b4ddbc0289b2e50d6de02d66a00c2b27.

svollebregt committed:
[P118] Adding Itho plugin to main reposity""

Adding P145 from the plugin playground as P118 to the main repo. Tested compiling of test_D and it seems to fit.

tonhuisman commented on src/_P118_Itho.ino:
```diff
@@ -0,0 +1,548 @@
+// code/idea was inspired by first release of code from 'Thinkpad'
+
+#include <SPI.h>
+#include "IthoCC1101.h"
```
Moving the library includes after the `#ifdef USES_P118` will eliminate the need to include SPI.h, and make the compiler job a tad lighter for the builds that don't have this plugin active.

svollebregt replied:
Done

tonhuisman commented on src/_P118_Itho.ino:
```diff
@@ -0,0 +1,548 @@
+
+#define PLUGIN_118
+#define PLUGIN_ID_118         118
+#define PLUGIN_NAME_118       "Itho ventilation remote"
```
All other plugins have a category included in their name. This could be `"Remote - Itho ventilation"` or something similar

svollebregt replied:
Changed

tonhuisman commented on src/_P118_Itho.ino:
```diff
@@ -0,0 +1,548 @@
+		{
+			addLog(LOG_LEVEL_DEBUG, F("UPDATE by PLUGIN_ONCE_A_SECOND"));
+			PLUGIN_118_Publishdata(event);
+			sendData(event);
```
If you use your own timer/interval mechanism, shouldn't the `Device[n].TimerOptional = true;` flag be added to the device flags? Now it will be triggered in 2 separate intervals.

svollebregt replied:
Added, I also disabled the timer option as it has no real use.

tonhuisman commented on src/_P118_Itho.ino:
```diff
@@ -0,0 +1,548 @@
+			if (cmd.equalsIgnoreCase(F("STATE")))
+			{
+
+				if (param1.equalsIgnoreCase(F("1111")))
```
String processing is quite 'expensive' on ESP's (and Arduinos), and as these values are all numeric, and ESPEasy already tries to provide 4 parsed arguments in `event->Par1..4`, you could refactor these `if`s into a `switch(event->Par1)...`

svollebregt replied:
Adjusted, now using Par1

svollebregt commented:
Thanks for all the comments @tonhuisman, I will incorporate the changes test and get back to this PR.

svollebregt committed:
Merge branch 'letscontrolit:mega' into mega

svollebregt committed:
Update _P118_Itho.ino

Incorporated and tested all changes suggested by tonhuisman:
- Moved ifdef to start
- Changed plugin name
- Adjusted timer options
- Switch statement using Par1

svollebregt committed:
Merge branch 'mega' of https://github.com/svollebregt/ESPEasy into mega

TD-er commented on src/_P118_Itho.ino:
```diff
@@ -87,7 +88,7 @@ volatile bool PLUGIN_118_Int = false;
 #define PLUGIN_118
 #define PLUGIN_ID_118         118
-#define PLUGIN_NAME_118       "Itho ventilation remote"
+#define PLUGIN_NAME_118       "Remote - Itho ventilation"
```
Doesn't really feel like the same style as other plugin names have.
Looks like there is a separate class of plugins called "remote", but I'm not sure if that really covers this. Does it?

svollebregt replied:
Perhaps Communication - Itho ventilation is better suited?

TD-er replied:
Yep, I think so.

svollebregt replied:
Implemented

TD-er commented on src/_P118_Itho.ino:
```diff
@@ -244,134 +246,149 @@ boolean Plugin_118(byte function, struct EventStruct *event, String &string)
+						success = true;
+						break;
+					}
+				 	case 9999:
```
Can you at least make defines for these "magical" numbers?
Makes it easier to read what you're doing here.

svollebregt replied:
They are explained at the start of the plugin, but I can add some local comments.

TD-er replied:
Comments tend to get outdated.
Why not use a `#define P118_BLADIEBLA 9999` and then replace all `9999` with this `P118_BLADIEBLA` ?

svollebregt replied:
I only use the numbers once and the numbers are set by the library and not bound to change. IMO a comment is a bit more straightforward.

TD-er commented on src/_P118_Itho.ino:
```diff
@@ -244,134 +246,149 @@ boolean Plugin_118(byte function, struct EventStruct *event, String &string)
+						interrupts();
+						PLUGIN_118_rf.initReceive();
+						addLog(LOG_LEVEL_INFO, F("Sent command for 'join' to Itho unit"));
+						printWebString += F("Sent command for 'join' to Itho unit");
```
You're doing this on every occurence where you add to the log and to the `printWebString`.
That's using the same code twice.
The compiler will do the deduplication, but it is harder to maintain to have the same string twice.
Can't you make a simple function for it?

svollebregt replied:
Implemented

TD-er commented on src/_P118_Itho.ino:
```diff
@@ -244,134 +246,149 @@ boolean Plugin_118(byte function, struct EventStruct *event, String &string)
+				 	case 9999:
+					{
+						noInterrupts();
+						PLUGIN_118_rf.sendCommand(IthoLeave);
```
Every call to `sendCommand` seems to be wrapped in `noInterrupts()` and `interrupts()`.
Doesn't it make more sense to have these in the sendCommand function?
Some occurences below do perform other assignments within the "scope" of these disabled interrupts.
Is it needed to do these with interrupts disabled?

Point is, it is easy to sometimes miss re-enabling the interrupt in your code, so I do want to see almost nowhere unless absolutely needed.
It has been the cause of a few very hard to reproduce bugs in the past, where a return was inserted in the code before the interrupts were enabled again.
So this is purely from a maintainance point of view.

svollebregt replied:
Good question if it is needed, I added this ~2 years ago because the plugin kept crashing. The source of this could have been somewhere else, though. The philosophy is that when we send we don't want an arriving RF packet to interrupt the sending, otherwise the actual state and send state might not be the same. The reason it is here and not in the library is that others also use the library in their projects and this makes it easier for me to update the library when changes are made.

I can disable the noInterrupts on my side and see if I notice any issues.

tonhuisman replied:
You could also add a `sendCommandNoInterrupt()` to the library that does just that, disable interrupt, send the command and re-enable the interrupt, killing 2 birds with 1 stone 😉

TD-er replied:
And since it is called several times, it will probably also reduce the binary size.

svollebregt replied:
I've moved the noInterrupts()/interrupts() code outside the switch statement similar as when receiving a RF packet. This will keep the library maintenance more straightforward (drop-in update save a single adjustment).

TD-er replied:
Nope, not a good idea, since you're writing to the log too, which may take "forever" and is very tricky to do without interrupts enabled.

Also "initreceive" was outside the scope of the disabled interrupts, so no idea what happens when that's initialized with interrupts disabled.

Disabling interrupts should be as short as possible (and not shorter ;)) and above all not longer than needed to keep the code more clean.
Better to find some way, like Ton mentioned, to have a function handling the interrupts and sending commands.

svollebregt replied:
I have moved the blocking code to the Itho library and also reduced it in the ITHOCheck routine that handles received RF packages.

svollebregt committed:
Merge branch 'letscontrolit:mega' into mega

svollebregt committed:
Update _P118_Itho.ino

Code clean-up following suggestions of TD-er.

svollebregt committed:
Merge branch 'mega' of https://github.com/svollebregt/ESPEasy into mega

TD-er committed:
Merge branch 'mega' into mega

svollebregt committed:
Merge branch 'letscontrolit:mega' into mega

svollebregt committed:
Merge branch 'letscontrolit:mega' into mega

svollebregt committed:
Further code improvements

-Moved noInterrupts() part to Itho library
- Reduced length of status log message
- Reduced noInterrupts() blocked code in ITHOCheck function (handles received RF packages) and made log more efficient

TD-er commented on lib/Itho/CC1101Packet.h:
```diff
@@ -0,0 +1,23 @@
+{
+	public:
+		uint8_t length;
+		uint8_t data[128];
```
Why is this packet not initialized?

```c++
uint8_t length = 0;
uint8_t data[128] = {0};
```

arjenhiemstra replied:
It should and looking through the code it indeed isn't. Probably something that has been in there from the first version of the lib (multiple devs ago... ;-))

TD-er commented on lib/Itho/IthoCC1101.cpp:
```diff
@@ -0,0 +1,852 @@
+#define DEBUG 0
+
+#define BYTE_TO_BINARY_PATTERN "%c,%c,%c,%c,%c,%c,%c,%c,"
+#define BYTE_TO_BINARY(byte)  \
```
`BYTE_TO_BINARY` and `BYTE_TO_BINARY_PATTERN` do not seem to be used.

arjenhiemstra replied:
This is a remnant of development debug code, can be removed

TD-er commented on lib/Itho/IthoCC1101.cpp:
```diff
@@ -0,0 +1,852 @@
+    finishTransfer();
+    delay(delaytime);
+  }
+  interrupts(); //Moved from plugin here to reduce blocking code
```
The scope where interrupts are disabled is quite long here and one of the biggest NO-NO's during interrupts is called: `delay`
There is a number of messages created. Why do you need to disable interrupts for that?
And sending data while interrupts are disabled is also not what I consider a good design.

Please have a good look at this. Why need interrupts be disabled here?

svollebregt replied:
I've tested this at my side and it appears these sections of interrupts disabled code are not necessary at all and are remnants of the original version of the plugin made by Jodur. I will test some more to see if it results somehow in instability or missed commands, but likely to be removed in the next commit.

I will check with arjenhiemstra who made the current version of the Itho library if the things you mentioned above about the library can be resolved.

tonhuisman commented on platformio.ini:
```diff
@@ -20,8 +20,8 @@ extra_configs =
-default_envs = custom_ESP8266_4M1M, custom_ESP32_4M316k
-; default_envs = custom_ESP8266_4M1M
+;default_envs = custom_ESP8266_4M1M, custom_ESP32_4M316k
+default_envs = test_D_ESP8266_4M1M
```
The `platformio.ini` file should not be committed unless there are required changes in there. This isn't required.

You'd better expand the desired PlatformIO environment (at least when using VSCode, don't know if Atom also supports that) and execute the Build action from there.

TD-er replied:
Yep Atom has a combobox with the PIO envs in the bottom left.

svollebregt replied:
My bad, I thought I removed this from the commit.

tonhuisman commented:
> Related to issue

If you change that text to 'Resolves' (this is a [helpful list on SO](https://stackoverflow.com/questions/1687262/link-to-the-issue-number-on-github-within-a-commit-message)) in the original comment, Github will link the PR to the Issue and close that when the PR is merged. Also, the Issue is then earmarked to have a resolving PR 😉. That'll keep the administration of issues nicely organized.

svollebregt committed:
Remove noInterrupts() and Itho library improvements

Related to #3801:
- Removed all noInterrupts(); code -> not needed and tested stable
- Initialize CC1101 packet
- Removed not used defines from IthoCC1101.ccp

TD-er committed:
Merge branch 'mega' into mega

svollebregt committed:
Merge branch 'mega' into mega

svollebregt committed:
Merge branch 'letscontrolit:mega' into mega

svollebregt committed:
Merge branch 'letscontrolit:mega' into mega

TD-er committed:
Merge branch 'mega' into mega

TD-er committed:
[PIO] Restore default envs
